File: JavaScript/TinyScript/src/parser/SimpleParser.js
const Expr = require('./ast/Expr')
const Scalar = require('./ast/Scalar')
const ASTNodeType = require('./ast/ASTNodeTypes')

class SimpleParser {

    /**
     * 转抽象语法树
     * @param {PeekTokenIterator | PeekIterator} it
     * @return {Expr | Scalar}
     */
    static parse(it) {

        // Expr -> digit + Expr | digit

        const expr = new Expr(null)
        const scalar = new Scalar(expr, it)

        if (!it.hasNext()) {
            return scalar
        }

        expr.addChild(scalar)
        const op = it.nextMatchOfValue('+')
        expr.label = '+'
        expr.type = ASTNodeType.BINARY_EXPR
        expr.lexeme = op
        expr.addChild(SimpleParser.parse(it))

        return expr
    }

}

module.exports = SimpleParser

File: Java/TinyScript/src/main/java/parser/SimpleParser.java
package parser;

import parser.ast.ASTNode;
import parser.ast.ASTNodeTypes;
import parser.ast.Expr;
import parser.ast.Scalar;
import parser.util.ParseException;
import parser.util.PeekTokenIterator;

public class SimpleParser {

    public static ASTNode parser(PeekTokenIterator it) throws ParseException {

        // Expr -> digit + Expr | digit

        var expr = new Expr(null);
        var scalar = new Scalar(expr, it);

        // base condition
        if (!it.hasNext()) return scalar;

        expr.setLexeme(it.peek());
        it.nextMatch("+");
        expr.setLabel("+");
        expr.addChild(scalar);
        expr.setType(ASTNodeTypes.BINARY_EXPR);
        var rightNode = parser(it);
        expr.addChild(rightNode);

        return expr;
    }

}
